### PR TITLE
Extend testing of corner cases of floating point parameter

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

This PR started as a follow up to a few deficiencies in testing that came up in https://github.com/cms-sw/cmssw/pull/45216. Following improvements are made
* Handling of `inf` and `nan` when generating the `cfi` files (in case anyone would use them as default values)
* Add checks for the max, `inf`, and `nan` for double and float to `ProducerWithPSetDesc`
* Extend `edmProvDump` tests to cover all parameter types with `ProducerWithPSetDesc`
  * The `V{Event,LuminosoityBlock}ID` and `V{Event,LuminosoityBlock}Range` printouts contained null characters, so the `operator<<()` of `edm::Entry` is improved to avoid those
  * `edmRNTupleTempProvDump` did not have any tests, so the `edmProvDump` tests are copied and converted to use the `RNTupleTemp` modules

Resolves https://github.com/cms-sw/framework-team/issues/2345

#### PR validation:

Unit tests pass

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@Dr15Jones Could you review?

File: FWCore/ParameterSet/src/Entry.cc [hunk 1]
```diff
@@ -1316,6 +1316,78 @@ namespace edm {
...
+        // encoded EventID
+        os << "{";
+        std::string_view start = "";
+        std::string_view const between(",");
+        std::string encoded;
+        for (auto const& id : entry.getVEventID()) {
+          // encode to have more compact presentation instead of EventID's own operator<<()
+          encode(encoded, id);
+          os << start << encoded;
+          encoded.clear();
```
Comment (Dr15Jones) [inline FWCore/ParameterSet/src/Entry.cc]: Is the hope that `clear` is faster than constructing/destructing `encoded` within the body of the loop?

File: FWCore/ParameterSet/src/Entry.cc [hunk 2]
```diff
@@ -1316,6 +1316,78 @@ namespace edm {
...
+        std::string_view const between(",");
+        std::string encoded;
+        for (auto const& range : entry.getVLuminosityBlockRange()) {
+          // encode to avoid quotation marks from LuminosityBlockRange's own operator<<()
+          encode(encoded, range);
+          os << start << encoded;
+          encoded.clear();
+          start = between;
+        }
+        os << "}";
```
Comment (Dr15Jones) [inline FWCore/ParameterSet/src/Entry.cc]: Given how all of these additions are basically identical, how about reusing a function (or templated function)?

File: FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc [hunk 3]
```diff
@@ -2033,6 +2033,34 @@ hadDefault = 'default'
     CHECK(edm::parameterTypeEnumToString(par->type()) == std::string("FileInPath"));
   }
 
+  SECTION("Corner cases of double") {
+    using Catch::Matchers::Equals;
+    SECTION("Infinity") {
```
Comment (Dr15Jones) [inline FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc:2062]: What about min, max and float?

File: FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc [hunk 4]
```diff
@@ -2033,6 +2033,34 @@ hadDefault = 'default'
...
+    SECTION("NaN") {
+      edm::ParameterSetDescription psetDesc;
+      psetDesc.add<double>("name", std::numeric_limits<double>::quiet_NaN());
+      std::ostringstream os;
+      edm::CfiOptions ops = edm::cfi::Typed{};
+      psetDesc.writeCfi(os, false, 0, ops);
+      REQUIRE_THAT(os.str(), Equals("\nname = cms.double(float('nan'))\n"));
+    }
+  }
+
```
Comment (Dr15Jones) [inline FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc:2087]: What about testing your changes to the various `V` types?

File: FWIO/RNTupleTempInput/test/run_TestEdmProvDump.sh [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    shift
+    cmsRun ${LOCAL_TEST_DIR}/${CONFIG} --version ${VERSION} $@ || die "cmsRun ${CONFIG} $@" $?
+}
+
+function compareProv {
+    OPTIONS=$1
+    FILE=$2
+    LOG=$3
+    edmRNTupleTempProvDump $OPTIONS $FILE | grep -v -E "$CHANGINGPARTS" > $LOG || die "edmProvDump $OPTIONS $FILE" $?
+    diff ${LOCAL_TEST_DIR}/unit_test_outputs/$LOG $LOG  || die "comparing $LOG" $?
```
Comment (Dr15Jones) [inline FWIO/RNTupleTempInput/test/run_TestEdmProvDump.sh:22]: Did you do a simple 'hand' test to check that if comparison fails the test fails?

**Review by Dr15Jones — CHANGES_REQUESTED** [review]

File: FWCore/ParameterSet/src/Entry.cc (hunk 1 above)
Comment (makortel) [reply FWCore/ParameterSet/src/Entry.cc]: Yes, because `clear()` doesn't necessarily deallocate.

File: FWCore/ParameterSet/src/Entry.cc (hunk 2 above)
Comment (makortel) [reply FWCore/ParameterSet/src/Entry.cc]: Thanks for the suggestion

File: FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc (hunk 3 above)
Comment (makortel) [reply FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc:2062]: Min and max didn't cause problems in cfi generation, but I can add them. Float I forgot, I'll add it.

File: FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc (hunk 4 above)
Comment (makortel) [reply FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc:2087]: The changes on the `V` types were in the `operator<<()` for `edm::Entry`, that is not related to the `ParameterSetDescription` facility.

The `operator<<()` is used by `edmProvDump`, but I'm not sure if is used anywhere else. It is not tested in any existing test in `FWCore/ParameterSet`. I _could_ add such a test, but we discussed earlier it might not be worth it and instead we'd rely on the `edmProvDump` tests.

File: FWIO/RNTupleTempInput/test/run_TestEdmProvDump.sh (hunk 5 above)
Comment (makortel) [reply FWIO/RNTupleTempInput/test/run_TestEdmProvDump.sh:22]: I did one now by commenting out my changes in `FWCore/ParameterSet/src/Entry.cc` and both this test and similar one in `IOPool/Common/test` report failures.

**Comment by makortel** [comment]
Addressed Chris' comments.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
Comparison differences are related to https://github.com/cms-sw/cmssw/issues/47071 and https://github.com/cms-sw/cmssw/issues/51594

**Comment by makortel** [comment]
+core

**Comment by mandrenguyen** [comment]
+1
